File: asn1jrt/src/main/java/org/asn1gen/runtime/java/TagClass.java
package org.asn1gen.runtime.java;

public enum TagClass {
  UNIVERSAL(0),
  APPLICATION(1),
  CONTEXT(2),
  PRIVATE(3);
  
  public final int value;
  
  TagClass(final int value) {
    this.value = value;
  }
  
  public static TagClass fromTagByte(final int value) {
    switch ((value >> 6) & 0x3) {
    case 0: return UNIVERSAL;
    case 1: return APPLICATION;
    case 2: return CONTEXT;
    case 3: return PRIVATE;
    default: throw new RuntimeException("invalid value");
    }
  }
}

File: asn1runtime/src/main/scala/org/asn1gen/runtime/codec/TagClass.scala
package org.asn1gen.runtime.codec

sealed abstract class TagClass(val value: Int) {
  val bits: Int = (value & 0x3) << 6 
}

object TagClass {
  object Universal extends TagClass(0)
  object Application extends TagClass(1)
  object ContextSpecific extends TagClass(2)
  object Private extends TagClass(3)
}
